text/0043-block-metadata.md

```diff
@@ -0,0 +1,193 @@
+
+    /// Quorum set configuration at the time of externalization.
+    QuorumSet quorum_set = 2;
+
+    /// IAS report for the enclave which generated the signature.
+    VerificationReport verification_report = 3;
+}
+
+message BlockMetadata {
+    /// Metadata signed by the consensus node.
+    BlockMetadataContents contents = 1;
+
+    /// Message signing key.
+    Ed25519Public node_key = 2;
+
+    /// Signature using `node_key` over the Digestible encoding of `contents`.
+    Ed25519Signature signature = 3;
+}
+
+// NB: Only addition is field #4.
+message ArchiveBlockV1 {
+    /// Block.
+    Block block = 1;
+
+    /// Contents of the block.
+    BlockContents block_contents = 2;
+
+    /// Block signature.
+    BlockSignature signature = 3;
+
+    /// Additional signed metadata about this block.
+    BlockMetadata metadata = 4;
+}
+```
+
+Since this is a protobuf-additive change, we don't need a new `ArchiveBlock`
+version/type.
+
+## Signing metadata
+The metadata signature will always differ from the block signature. The block
+signature is generated by the enclave using a private key that never leaves the
+enclave, while the metadata is signed by the "untrusted" code outside the
+enclave using the SCP message signing key, since the enclave does not have this
+metadata.
+
+The metadata signature is generated from the `Digestible` encoding of
+`BlockMetadataContents`, since the protobuf wire format is not a canonical
+encoding.
+
+## Signing key history
+MobileCoin Foundation will publish a mapping/list of node_id to a range of block
+indexes, which node operators and consumers can use to verify the metadata
+signatures.
+
+This mapping will contain multiple node_id:block_range entries, in the following
+TOML format:
+```toml
+[[node]]
+message_signing_pub_key = ""  # Pub key, as PEM-encoded data or a file path containing such data.
+first_block_index = X  # Index of the first block signed with this key.
+last_block_index = Y  # Optional index of the last block signed with this key.
+```
+
+## Historical AVRs
+With `BlockMetadata`, consumers of the blockchain no longer need a Watcher DB.
+Ledger DB will default to reading AVRs from `BlockMetadata`, and fall back to a
+lookup table containing AVRs for older blocks.
+
+MobileCoin Foundation will publish a bootstrap file defining this lookup table
+with historical consensus enclave AVRs, up to when consensus nodes publish
+`BlockMetadata` with their AVRs.
+
+The lookup data will be in the following TOML format:
+```toml
+[[node]]
+responder_id = ""  # Responder ID for the consensus node.
+avr = ""  # AVR for the node's enclave, includes pubkey.
+first_block_index = X  # Index of the first block signed with this enclave.
+last_block_index = Y  # Index of the last block signed with this enclave.
+```
+
+## Block Version bump
+Block version 3 (release v1.3) will make the `block_metadata` required, at which
+point consensus nodes and clients should reject blocks without metadata.
```

Comment:
How does work when a node is performing catch up (for example when syncing blocks that were formed when it was down). Does it arbitrarily picks up metadata from one of the sources it syncs from?

Reply:
The `ArchiveBlocks` in S3 already include the `BlockSignature` from which watcher derives the block timestamp. AFAICT that is sufficient for catching up.

Reply:
Unless the code changed, the [ledger sync code](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/ledger/sync/src/ledger_sync/ledger_sync_service.rs#L211) does not append the signature. It doesn't do that since the node has no way of validating that this signature actually came from a legitimate node ( in theory someone could replace the s3 contents with whatever they want right now). We made the decision that a node's local ledger only contains signature generated by the node itself, which also means you do not end up with a ledger that has a mishmash of signatures from multiple nodes.

Reply:
The AVR + signing key validation is integral for ensuring the metadata comes from a trusted consensus node, and since the metadata includes the AVR, we can validate the enclave key used for the block signature, and thus authenticate the signature.

> Unless the code changed, the [ledger sync code](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/ledger/sync/src/ledger_sync/ledger_sync_service.rs#L211) does not append the signature.

I switched to adding the whole `BlockData` in https://github.com/mobilecoinfoundation/mobilecoin/pull/2052; it sounds like that should not be merged until the AVR + signing key validation is.

Reply:
That will only work going forward.

Right now it is possible there are blocks in S3 what contain no signature (e.g. if a node performed catchup at any point, it will then upload signature-less blocks into its S3 bucket).
Looking at https://github.com/mobilecoinfoundation/mobilecoin/blob/block-metadata-in-ledger/ledger/sync/src/ledger_sync/ledger_sync_service.rs, unless I am missing something, I am not seeing a path that prioritizes blocks that do we a signature and timestamp.

This MCIP is not necessarily the right place to continue this discussion, but my point is we should be careful about hinting the watcher can be removed. Unless the ledger syncing code changes to download a block from multiple sources until a version of it with a signature/timestamp is located, the watcher is still needed.

Reply:
I see. Thank you for flagging!

Reply:
To close the loop on this, https://github.com/mobilecoinfoundation/mobilecoin/pull/2216 handles catch-up mode